File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;
import java.util.stream.*;

class Solution {
    
    private List<Edge>[] graph;
    private int[] dists;
    private boolean[] visited;
    
    static class Edge implements Comparable<Edge>{
        int node;
        int weight;
        
        public Edge(int node, int weight) {
            this.node = node;
            this.weight = weight;
        }
        
        @Override
        public int compareTo(Edge e) {
            return this.weight - e.weight;
        }
    }

    public int solution(int n, int[][] edge) {
        int answer = 0;
        graph = new ArrayList[n + 1];
        dists = new int[n + 1];
        visited = new boolean[n + 1];
        
        for (int i = 1; i <= n; i++) {
            graph[i] = new ArrayList<>();
        }
        
        Arrays.fill(dists, Integer.MAX_VALUE);
        
        for (int[] e : edge) {
            int start = e[0];
            int end = e[1];
            graph[start].add(new Edge(end, 1));
            graph[end].add(new Edge(start, 1));
        }
        
        dijkstra();
        
        int max = Arrays.stream(dists).skip(1).max().orElse(Integer.MIN_VALUE);
        answer = (int) Arrays.stream(dists).filter(num -> num == max).count();
        
        return answer;
    }
    
    private void dijkstra() {
        
        PriorityQueue<Edge> queue = new PriorityQueue<>();
        queue.offer(new Edge(1,0));
        dists[1] = 0;
        
        while (!queue.isEmpty()) {
            Edge edge = queue.poll();
            int node = edge.node;
            
            if (visited[node]) {
                continue;
            }
            visited[node] = true;
            
            for (Edge e : graph[node]) {
                if (dists[e.node] > dists[node] + e.weight) {
                    dists[e.node] = dists[node] + e.weight;
                    queue.offer(new Edge(e.node, dists[e.node]));
                }
            }
        }
    }
    
    
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
import heapq

def solution(n, edge):
    graph = []
    for i in range(n + 1):
        graph.append([])

    for start, end in edge:
       graph[start].append([1, end]) 
       graph[end].append([1, start])

    dists = [float('inf')] * (n + 1)
    dists[1] = 0
    visited = [False] * (n + 1)
    heap = [[0, 1]]
    
    while heap:
        cost, node = heapq.heappop(heap)
        if (visited[node]):
            continue
        visited[node] = True

        for nxt_cost, nxt_node in graph[node]:
            new_cost = dists[node] + nxt_cost
            if dists[nxt_node] > new_cost:
                dists[nxt_node] = new_cost
                heapq.heappush(heap, [new_cost, nxt_node])
    
    return dists.count(max(dists[1:]))